teuthology/lock/ops.py

```diff
@@ -297,3 +285,153 @@ def push_new_keys(keys_dict, reference):
                 log.error('failed to update %s!', hostname)
                 ret = 1
     return ret
+
+
+def reimage(ctx, machines, machine_type):
+    reimaged = dict()
+    with teuthology.parallel.parallel() as p:
+        for machine in machines:
+            log.info("Start node '%s' reimaging", machine)
+            update_nodes([machine], True)
+            p.spawn(teuthology.provision.reimage, ctx,
+                    machine, machine_type)
+            reimaged[machine] = machines[machine]
+            log.info("Node '%s' reimaging is complete", machine)
+    return reimaged
+
+
+def reimage_many(ctx, machines, machine_type):
+    # Setup log file, reimage machines and update their keys
+    reimaged = dict()
+    console_log_conf = dict(
+        logfile_name='{shortname}_reimage.log',
+        remotes=[teuthology.orchestra.remote.Remote(machine)
+                 for machine in machines],
+    )
+    with console_log.task(ctx, console_log_conf):
+        reimaged = reimage(ctx, machines, machine_type)
+    reimaged = do_update_keys(list(reimaged.keys()))[1]
+    update_nodes(reimaged)
+    return reimaged
+
+
+def block_and_lock_machines(ctx, total_requested, machine_type, reimage=True):
```

Comment:
Why this so strange name? block_and_lock_machines? what doe it mean block?

Reply:
I suppose you meant that lock_machines will be trying to lock in loop even if it was not able to actually lock or provision the node successfully.
This is probably out of the scope for this PR but this method should be combined in future with lock_many, and lock_one, and present the same method which is used by: dispatcher, the task, and lock cli.

Reply:
Yeah, I used `block_and_lock_machines` since we are initially waiting to block machines and then lock them. 

Like you said, the locking mechanism is definitely split in a lot of places and could use some refactoring.